packages/types/lib/ExpressionEvaluator.ts

```diff
@@ -57,11 +65,56 @@ export interface IExpressionEvaluator {
    */
   evaluateAsEBV: (mapping: RDF.Bindings) => Promise<boolean>;
 
+  evaluateAsInternal: (mapping: RDF.Bindings) => Promise<E.Expression>;
+}
+
+export interface IOrderByEvaluator {
   /**
    * Orders two RDF terms according to: https://www.w3.org/TR/sparql11-query/#modOrderBy
    * @param termA the first term
    * @param termB the second term
-   * @param strict whether to throw an error (true), or compare by value (false) if no other compare rules match.
    */
-  orderTypes: (termA: RDF.Term | undefined, termB: RDF.Term | undefined, strict: boolean | undefined) => -1 | 0 | 1;
+  orderTypes: (termA: RDF.Term | undefined, termB: RDF.Term | undefined) => -1 | 0 | 1;
+}
+
+export interface IEvalContext {
+  args: E.Expression[];
+  mapping: RDF.Bindings;
+  exprEval: ContextualizedEvaluator;
+}
+
+export type FunctionApplication = (evalContext: IEvalContext) => Promise<E.TermExpression>;
+
+export type FunctionExpression = IExpressionFunction | ITermFunction;
+
+export interface IExpressionFunction {
+  definitionType: 'onExpression';
+  apply: (evalContext: IEvalContext) => Promise<E.TermExpression>;
+  /**
+   * Makes you able to error in the termTransformer.
+   */
+  checkArity: (args: E.Expression[]) => boolean;
+}
+
+export interface ITermFunction {
```

Comment:
Since an `ITermFunction` also appears to be an `IExpressionFunction`, perhaps we can let the former extend the latter?
The latter could then just have a boolean field `supportsTermExpressions: true`.